Add inverse for Tensor

fmahebert opened the pull request:
Questions:

1. In the code call we leaned towards only implementing 3x3. Do we still agree? My concern is that we may find a good reason to use a 4x4 inverse in the future but forget how to do it correctly. Should we include Saul's efficient 4x4 algorithm somewhere in the repo for this case?

2. I would like to rename the file `InvertMatrix` to `Inverse` so that it matches the function it defines. However, this does make the name more generic / less descriptive. Opinions?

fmahebert committed:
Add determinant_and_inverse and associated test

fmahebert commented on tests/Unit/Numerical/LinearAlgebra/Test_InvertMatrix.cpp:
```diff
@@ -0,0 +1,172 @@
+                      SpatialIndex<Dim, UpLo::Up, Frame::Inertial>>>;
+
+// Checks that the inverse of a Tensor<double, Symm, index_list<I1, I2>>
+// is a Tensor<double, Symm, index_list<I3, I4>.
```
This would be cleaner if it used `change_index_up_lo`, just as in the `inverse` implementation. I opted to manually control the indices on the left/right hand sides so that the test stays more orthogonal to the code it is testing.

kidder replied:
This seems unnecessarily complicated to me; why can't I just one-line it by  (figuratively)
auto inv = ...
static_assert(decltype(inv) == what I expect)

fmahebert replied:
Here's what I could replace it with:
```
auto inv = inverse(orig_t{}, det_t{});
static_assert(cpp17::is_same_v<inv_t, decltype(inv)>, "");
```

This reduces the complexity associated with getting the function's return type, but introduces the complexity of creating a temporary object. I have no strong preference.

kidder replied:
sorry, I wasn't clear;  can't you just do the one-line static assert after you have created the objects below in the other tests (i.e. after lines 47, 58, 77, 95, 111 and delete this function and lines 127-139

fmahebert replied:
Yes I could do this.. but I'm not sure I understand the motivation. Testing that the inverse Tensor has the expected indices requires exposing the Tensor template parameters, which makes the associated interfaces more complex. Instead of having some complex-looking code in this function and in lines 127-139, the complex-looking code would be in every `verify_inverse_X` function and its call sites.

I will write this as a fixup commit and we can see how we feel about it.

kidder commented:
1) I agree with Francois, InvertMatrix -> Inverse.
2) Is Saul's 4d fast matrix inversion just valid for the spacetime metric, or any arbitrary tensor?
3) General question for the audience: why are Inverse and Determinant in LinearAlgebra and not Tensor?  I would expect functions named inverse and determinant in a LinearAlgebra directory to be able to take the inverse and determinant of arbitrary size square matrices...

kidder commented on tests/Unit/Numerical/LinearAlgebra/Test_InvertMatrix.cpp:
```diff
@@ -0,0 +1,172 @@
+using symmetric = tmpl::integral_list<int32_t, 1, 1>;
+
+template <size_t Dim>
+using matrix_ii = Tensor<double, symmetric,
```
I would like to avoid the word matrix as Matrix is used for run-time matrices in SpECTRE and Tensors are compile-time and limited to 4-dimensions

kidder replied:
Also why aren't you using the type aliases in DataStructures/Tensor/TypeAliases.hpp ?

fmahebert replied:
Because I wanted a Tensor with indices in 2 frames (Grid, Inertial). I agree I could use the type aliases for the other cases.

fmahebert commented:
1. ✅
2. It is an algorithm for inverting 4x4 matrices (in the math sense.. no physics involved).
3. This was initially for lack of a better suggestion. In #98 (main discussion, not under any specific code) @nils and I considered moving Determinant and Inverse to a different location, perhaps `Tensor/EagerMath`. If we come to a consensus, I am happy to move these files.

nilsdeppe reviewed (changes requested):
:+1: Rebase and squash

fmahebert commented:
We still need to settle two questions:

1. Whether to implement 4x4 inverses, or find some other way of saving Saul's algorithm in case of future need.
2. Which directory the Tensor inverse (and determinant) functions should live in.

nilsdeppe commented:
1. (After group discussion) Add it, but add a comment to the doxygen documentation of inverse that it is more efficient to use the lapse, shift, spatial metric approach Mark wrote if all three are also needed.
2. `src/DataStructures/Tensor/EagerMath`

fmahebert commented:
When I move `determinant` and `inverse` into `Tensor/EagerMath`, we may want to reconsider the location of the corresponding tests. I think it’s surprising for `tests/Unit/DataStructures/` to contain tests named `DivideBy`, `Determinant`, `Inverse`, etc, alongside the tests of actual data structures (which is what I expect this dir to hold).

I can think of two options:

1. move these tests into a subdir, e.g. `tests/Unit/DataStructures/TensorEagerMath`, or
2. rename these tests so their association with the Tensor class is made clear, e.g. `Test_Tensor<Op>.cpp`.

I prefer (1), as the organization is cleaner, but either sounds reasonable to me. Thoughts?

nilsdeppe commented:
1 sounds good to me

kidder commented:
rebase on Travis fixes

fmahebert committed:
Stylistic update to determinant code and test

Reorder terms to make `determinant` and `determinant_and_inverse` share
the same convention for symmetric tensors, i.e. explicitly handle the
above-diagonal terms.

No functionality change.

fmahebert commented:
I've updated this PR with substantial changes:

1. The branch is rebased and updated
4. The function `inverse` is replaced with `determinant_and_inverse` which computes both quantities together for efficiency
2. The determinant/inverse code is moved to `Tensor/EagerMath`
3. The 4x4 tensor inverse is now implemented

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/MakeArithmeticValue.hpp"
+
+/*
```
If you just use a pair instead of this class we could actually use `std::tie` and structured bindings in C++17.

fmahebert replied:
Edit: Comment removed because I misunderstood Nils' intent.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+      inv;
+};
+
+namespace detail {
```
I prefer `thing_its_used_in_detail` or `ThingItsUsedIn_detail` to avoid unintentional leakage into different detail namespaces.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+namespace detail {
+// Helps to shorten some repeated code:
+template <typename Index0, typename Index1>
+using InverseIndices =
```
metafunctions are snake_case

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+
+template <typename Symm, typename Index0, typename Index1>
+struct DetAndInverseImpl<Symm, Index0, Index1,
+                         std::enable_if_t<Index0::dim == 1>> {
```
use `Requires` instead of `enable_if_t`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+  static TensorDetAndInverse<T, Symm, Index0, Index1> apply(
+      const Tensor<T, Symm, typelist<Index0, Index1>>& tensor) {
+    const T& t00 = tensor.template get<0, 0>();
+    const Scalar<T> det{t00};
```
What you're doing here is copying `t00` into the Scalar, then later in the return you copy the `Scalar` into the return type, so you are copying twice here. You should have just `{t00, inv}`, but since this will return a pair in the future it'll be `make_pair` instead of `{}`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    const T& t01 = tensor.template get<0, 1>();
+    const T& t10 = tensor.template get<1, 0>();
+    const T& t11 = tensor.template get<1, 1>();
+    const Scalar<T> det{t00 * t11 - t01 * t10};
```
Don't have this be const so we can use RVO and move semantics

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    const T& t00 = tensor.template get<0, 0>();
+    const T& t01 = tensor.template get<0, 1>();
+    const T& t11 = tensor.template get<1, 1>();
+    const Scalar<T> det{t00 * t11 - t01 * t01};
```
non const

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    const T a = t11 * t22 - t12 * t21;
+    const T b = t12 * t20 - t10 * t22;
+    const T c = t10 * t21 - t11 * t20;
+    const Scalar<T> det{t00 * a + t01 * b + t02 * c};
```
non-const

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    const T a = t11 * t22 - t12 * t12;
+    const T b = t12 * t02 - t01 * t22;
+    const T c = t01 * t12 - t11 * t02;
+    const Scalar<T> det{t00 * a + t01 * b + t02 * c};
```
non-const

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+//   W X
+// Each inverse block is obtained from the blocks {P,Q,R,S} of the input matrix
+// as follows:
+//   U = invP + invP Q X R invP
```
Have these be in the order they're implemented in below, and actually what is implemented below.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    U10 = invP10 - (invPQ10 * W00 + invPQ11 * W10);
+    U11 = invP11 - (invPQ10 * W01 + invPQ11 * W11);
+
+    const Scalar<T> det{detP * det_invX};
```
non-const

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+  template <typename T>
+  static TensorDetAndInverse<T, Symmetry<2, 1>, Index0, Index1> apply(
+      const Tensor<T, Symmetry<2, 1>, typelist<Index0, Index1>>& tensor) {
+    const T& P00 = tensor.template get<0, 0>();
```
snake_case (small `p`) There is not consensus on this, so it's up to you.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    T& X10 = inv.template get<3, 2>();
+    T& X11 = inv.template get<3, 3>();
+
+    const T detP = P00 * P11 - P01 * P10;
```
`det_p` and `inv_p`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+    const T invP10 = -P10 * one_over_detP;
+    const T invP11 = P00 * one_over_detP;
+
+    const T RinvP00 = R00 * invP00 + R01 * invP10;
```
`r_inv_p00`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+};
+
+template <typename Index0>
+struct DetAndInverseImpl<Symmetry<1, 1>, Index0, Index0,
```
do same stuff that makes sense :)

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+}  // namespace detail
+
+/*!
+ * \ingroup Functors
```
`NumericalAlgorithms`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+ * \ingroup Functors
+ * \brief Computes the determinant and inverse of a rank-2 tensor.
+ *
+ * Computes and the determinant and inverse together, because this leads to
```
remove first `and`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+ * \brief Computes the determinant and inverse of a rank-2 tensor.
+ *
+ * Computes and the determinant and inverse together, because this leads to
+ * fewer operations as compared to computing the determinant independently.
```
Please remove `as`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+ * fewer operations as compared to computing the determinant independently.
+ *
+ * \param tensor the input rank-2 Tensor.
+ * \return a `TensorDetAndInverse` struct that holds both the determinant and
```
update this to `pair`

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+ *
+ * \note
+ * When inverting a 4x4 spacetime metric, it is typically more efficient to use
+ * the ADM representation of the 4-metric in terms of lapse, shift, and spatial
```
"3+1 decomposition" instead of "ADM representation"

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,357 @@
+  static_assert(Index0::index_type == Index1::index_type,
+                "Taking the inverse of a mixed Spatial and Spacetime index "
+                "Tensor is not allowed since it's not clear what that means.");
+  static_assert(not tt::is_integer_v<T>, "Can't invert a Tensor<int>.");
```
Use `std::is_integral` since we probably shouldn't invert matrices of chars either ;)

nilsdeppe commented on tests/Unit/DataStructures/TensorEagerMath/Test_Determinant.cpp:
```diff
@@ -22,8 +22,8 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.EagerMath.Determinant",
     {
       tnsr::ij<double, 2, Frame::Grid> matrix{};
       matrix.get<0, 0>() = 2.1;
```
We're confused with why the tests are changing. Changing the implementation shouldn't require the tests to be changed, so this does not fill us with confidence that the changed implementation is correct.

fmahebert replied:
Changing the implementation did not require the test to change :)

The change to the test is also stylistic and for consistency with `determinant_and_inverse`: I am switching from a "column major" style ordering where the first index varies fastest, to a "row major" style ordering where the last index varies fastest. It's true that I swapped indices rather than reordering lines of code, and therefore the matrix being passed into the `determinant` function got transposed. However, as expected, its determinant doesn't change (within roundoff). This change to the test is independent from the change to the implementation.

If it gives you increased confidence, I can change this commit to reorder lines of code instead of swapping indices: this way the matrix will be identical, and only the way it is displayed will change.

nilsdeppe commented on tests/Unit/DataStructures/TensorEagerMath/Test_DeterminantAndInverse.cpp:
```diff
@@ -0,0 +1,249 @@
+// In the spirit of the Tensor type aliases, but for a rank-2 Tensor with each
+// index in a different frame. If Fr1 == Fr2, then this reduces to tnsr::iJ.
+template <typename DataType, size_t Dim, typename Fr1, typename Fr2>
+using tnsr__iJ = Tensor<DataType, tmpl::integral_list<int32_t, 2, 1>,
```
This looks like a double `_`, which is UB, the standard does not allow it. Please use just one `_`

fmahebert replied:
I was using that for visual alignment/consistency with `tnsr::ij`, forgot it's reserved :(

nilsdeppe commented on tests/Unit/DataStructures/TensorEagerMath/Test_DeterminantAndInverse.cpp:
```diff
@@ -0,0 +1,249 @@
+  t.template get<2, 3>() = 13.0;
+  t.template get<3, 3>() = 17.0;
+  const auto result = determinant_and_inverse(t);
+  CHECK(result.det == determinant(t));
```
Should hard-code the det's as well (above too) and below forr

nilsdeppe commented on tests/Unit/DataStructures/TensorEagerMath/Test_DeterminantAndInverse.cpp:
```diff
@@ -0,0 +1,249 @@
+    using ia = SpatialIndex<2, UpLo::Lo, Frame::Grid>;
+    using iB = SpatialIndex<2, UpLo::Up, Frame::Inertial>;
+    using ib = SpatialIndex<2, UpLo::Lo, Frame::Inertial>;
+    Tensor<double, generic, index_list<iA, ib>> t{};
```
Please see if you can simplify this with `decltype`

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+    // inv is non-const so that it can be moved into the std::pair:
+    Tensor<T, Symm, inverse_indices<Index0, Index1>> inv{
+        make_arithmetic_value(t00, 1.0) / t00};
+    return std::make_pair(Scalar<T>{t00}, inv);
```
`std:move(inv)`

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+    inv.template get<0, 1>() = -t01 * one_over_det;
+    inv.template get<1, 0>() = -t10 * one_over_det;
+    inv.template get<1, 1>() = t00 * one_over_det;
+    return std::make_pair(det, inv);
```
`std::move(det), std::move(inv)`

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+    inv.template get<0, 0>() = t11 * one_over_det;
+    inv.template get<0, 1>() = -t01 * one_over_det;
+    inv.template get<1, 1>() = t00 * one_over_det;
+    return std::make_pair(det, inv);
```
moves

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+    inv.template get<2, 0>() = c * one_over_det;
+    inv.template get<2, 1>() = (t20 * t01 - t21 * t00) * one_over_det;
+    inv.template get<2, 2>() = (t00 * t11 - t01 * t10) * one_over_det;
+    return std::make_pair(det, inv);
```
moves

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+    inv.template get<1, 1>() = (t22 * t00 - t02 * t02) * one_over_det;
+    inv.template get<1, 2>() = (t02 * t01 - t00 * t12) * one_over_det;
+    inv.template get<2, 2>() = (t00 * t11 - t01 * t01) * one_over_det;
+    return std::make_pair(det, inv);
```
moves

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+
+    // det is non-const so that it can be moved into the std::pair:
+    Scalar<T> det{det_p * det_inv_x};
+    return std::make_pair(det, inv);
```
moves

wthrowe commented on src/DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp:
```diff
@@ -0,0 +1,365 @@
+
+    // det is non-const so that it can be moved into the std::pair:
+    Scalar<T> det{det_p * det_inv_x};
+    return std::make_pair(det, inv);
```
moves

nilsdeppe commented on tests/Unit/DataStructures/TensorEagerMath/Test_DeterminantAndInverse.cpp:
```diff
@@ -0,0 +1,248 @@
+    using result_t = std::result_of_t<function_t(input_t)>;
+    using inv_t = Tensor<double, generic, index_list<iB, ia>>;
+    static_assert(cpp17::is_same_v<inv_t, result_t::second_type>,
+                  "Inverse tensor has incorrect index structure.");
```
Can you try this beauty out:

```cpp
static_assert(
    cpp17::is_same_v<
        Tensor<double, tmpl::integral_list<int32_t, 2, 1>,
               index_list<SpatialIndex<2, UpLo::Up, Frame::Inertial>,
                          SpatialIndex<2, UpLo::Lo, Frame::Grid>>>,
        decltype(determinant_and_inverse(
            std::declval<Tensor<
                double, tmpl::integral_list<int32_t, 2, 1>,
                index_list<SpatialIndex<2, UpLo::Up, Frame::Grid>,
                           SpatialIndex<2, UpLo::Lo, Frame::Inertial>>>>()))>,
    "Inverse tensor has incorrect index structure.");
```

fmahebert replied:
This _is_ the droid I was looking for (modulo `::second_type`).

kidder reviewed (changes requested):
Looks good; squash please :1st_place_medal:

fmahebert commented:
Rebased and squashed
